crates/turbopath/src/relative_unix_path_buf.rs

```diff
@@ -1,75 +1,163 @@
-use std::path::PathBuf;
+use std::{fmt::Debug, io::Write};
 
-use serde::Serialize;
+use bstr::{BString, ByteSlice};
 
-use crate::{IntoUnix, PathValidationError};
+use crate::{not_relative_error, PathError, PathValidationError};
 
-#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default, Serialize)]
-pub struct RelativeUnixPathBuf(PathBuf);
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct RelativeUnixPathBuf(BString);
 
 impl RelativeUnixPathBuf {
-    /// Create a new RelativeUnixPathBuf from a PathBuf by calling `into_unix()`
-    ///
-    /// NOTE: `into_unix` *only* converts Windows paths to Unix paths *on* a
-    /// Windows system. Do not pass a Windows path on a Unix system and
-    /// assume it'll be converted.
-    ///
-    /// # Arguments
-    ///
-    /// * `path`:
-    ///
-    /// returns: Result<RelativeUnixPathBuf, PathValidationError>
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// ```
-    pub fn new(path: impl Into<PathBuf>) -> Result<Self, PathValidationError> {
-        let path = path.into();
-        if path.is_absolute() {
-            return Err(PathValidationError::NotRelative(path));
+    pub fn new(path: impl Into<Vec<u8>>) -> Result<Self, PathError> {
+        let bytes: Vec<u8> = path.into();
+        if !bytes.is_empty() && bytes[0] == b'/' {
+            return Err(not_relative_error(&bytes).into());
         }
+        Ok(Self(BString::new(bytes)))
+    }
+
+    pub fn as_str(&self) -> Result<&str, PathError> {
+        let s = self.0.to_str()?;
+        Ok(s)
+    }
+
+    // write_escaped_bytes writes this path to the given writer in the form
+    // "<escaped path>", where escaped_path is the path with '"' and '\n'
+    // characters escaped with '\'.
+    pub fn write_escapted_bytes<W: Write>(&self, writer: &mut W) -> Result<(), PathError> {
+        writer.write_all(&[b'\"'])?;
+        let mut i: usize = 0;
+        while i < self.0.len() {
+            if let Some(mut index) = self.0[i..]
```

Comment:
If we're doing this in a loop, isn't this O(n^2)? Could we loop through the slice and upon encountering `\"`, we write a `\\` ahead of it?

Reply:
`i` is the index into `self.0`. Each time we encounter a byte that needs escaping (@ `index`), we write out everything from `i` to `index`, then  `\\` and the byte-to-be-escaped, then increment `i` by one + `index`, moving us past the byte we just escaped. I'll add a comment.